### PR TITLE
Add indeterminate checkbox state and groupable checkboxes

### PR DESCRIPTION
**Pull request by pochretien**

### CODE REVIEW

File: src/components/checkbox/GroupableCheckboxConnected.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  parentId?: string;
+}
+
+const mapStateToProps = (state: IReactVaporState, ownProps: IGroupableCheckboxOwnProps): IInputStateProps => {
+  const groupableCheckboxesState: IGroupableCheckboxesState = _.findWhere(state.groupableCheckboxes, { parentId: ownProps.parentId || ownProps.id });
+  if (groupableCheckboxesState) {
+    const checkbox: ICheckboxState = ownProps.isParent ? groupableCheckboxesState.parent : _.findWhere(groupableCheckboxesState.checkboxes, {id: ownProps.id});
+    return {
+      checked: checkbox && checkbox.checked,
+      indeterminate: ownProps.isParent && groupableCheckboxesState.nbChecked > 0 &&
```
Comment (sroy3) [inline src/components/checkbox/GroupableCheckboxConnected.tsx]: Each condition should be on its own line

File: src/components/checkbox/GroupableCheckboxReducers.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+export const groupableCheckboxesInitialState: IGroupableCheckboxesState[] = [];
+
+export const groupableCheckboxReducer = (state: IGroupableCheckboxesState = groupableCheckboxInitialState,
+                                         action: IReduxAction<IGroupableCheckboxActionPayload>): IGroupableCheckboxesState => {
+  switch (action.type) {
+    case GroupableCheckboxActions.addGroup:
+      if (action.payload.isParent) {
+        return addParentCheckbox(state, action);
+      }
+      return addChildCheckbox(state, action);
```
Comment (sroy3) [inline src/components/checkbox/GroupableCheckboxReducers.ts]: This could be a ternary

File: src/components/checkbox/GroupableCheckboxReducers.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            parent: {
+              id: action.payload.parentId,
+              checked: nbChecked === state.total,
+            },
+            nbChecked,
+            checkboxes: [...checkboxes],
+          };
+        }
+      }
+      return state;
```
Comment (sroy3) [inline src/components/checkbox/GroupableCheckboxReducers.ts]: You could omit the return here and it would fall into the default case, returning the same thing.

File: src/components/checkbox/GroupableCheckboxReducers.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            ...state,
+            parent: {
+              id: action.payload.parentId,
+              checked: nbChecked === state.total,
+            },
+            nbChecked,
+            checkboxes: [...checkboxes],
+          };
+        }
+      }
```
Comment (sroy3) [inline src/components/checkbox/GroupableCheckboxReducers.ts]: We use switch/case in reducers because it is easy to see what is being returned. Here, things are getting a little blurry. Maybe you can simplify but moving the inside of your if statements in functions or something similar.

File: src/components/checkbox/GroupableCheckboxReducers.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+export const groupableCheckboxesReducer = (state: IGroupableCheckboxesState[] = groupableCheckboxesInitialState,
+                                           action: IReduxAction<IGroupableCheckboxActionPayload>): IGroupableCheckboxesState[] => {
+  let groupableCheckboxesState: IGroupableCheckboxesState;
+  switch (action.type) {
+    case GroupableCheckboxActions.addGroup:
+      groupableCheckboxesState = _.findWhere(state, { parentId: (action.payload.parentId || action.payload.id) });
+      if (state && groupableCheckboxesState) {
+        groupableCheckboxReducer(groupableCheckboxesState, action);
+        return state;
+      }
```
Comment (sroy3) [inline src/components/checkbox/GroupableCheckboxReducers.ts]: Why are you returning the state here?

File: src/components/checkbox/GroupableCheckboxReducersUtils.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+  if (state === groupableCheckboxInitialState) {
+    return {
+      ...state,
+      parentId: action.payload.id,
+      parent: newCheckbox,
+    };
+  }
+
+  state.parent = newCheckbox;
```
Comment (sroy3) [inline src/components/checkbox/GroupableCheckboxReducersUtils.ts]: You should not directly change the state. You have to send a copy of it. So here, you should return 
```
return {
  ...state,
  parent: newCheckbox,
};
```

File: src/components/checkbox/GroupableCheckboxReducersUtils.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    return {
+      ...state,
+      parentId: action.payload.parentId,
+      total,
+      nbChecked,
+      checkboxes: newChecboxes,
+    };
+  }
+
+  return _.extend(state, {
```
Comment (sroy3) [inline src/components/checkbox/GroupableCheckboxReducersUtils.ts]: You have to extend in an empty object so you do not change the state object.

File: src/components/checkbox/GroupableCheckboxReducersUtils.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    total,
+    nbChecked,
+  });
+};
+
+export const removeParentCheckbox = (state: IGroupableCheckboxesState[],
+                                     action: IReduxAction<IGroupableCheckboxActionPayload>): IGroupableCheckboxesState[] => {
+  return _.reject(state, (checkboxState: IGroupableCheckboxesState) => {
+    return action.payload.id === checkboxState.parentId;
+  });
```
Comment (sroy3) [inline src/components/checkbox/GroupableCheckboxReducersUtils.ts]: You can omit the return statement and curly braces `return _.reject(state, (checkboxState: IGroupableCheckboxesState) => action.payload.id === checkboxState.parentId);`

File: src/components/checkbox/GroupableCheckboxReducersUtils.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+};
+
+export const removeChildCheckbox = (state: IGroupableCheckboxesState,
+                                    action: IReduxAction<IGroupableCheckboxActionPayload>) => {
+  const checkboxState: ICheckboxState = _.findWhere(state.checkboxes, {id: action.payload.id});
+  if (checkboxState) {
+    state.checkboxes = _.reject(state.checkboxes, (checkbox: ICheckboxState) => checkbox.id === action.payload.id);
+    state.nbChecked += (checkboxState && checkboxState.checked ? -1 : 0);
+    state.total += checkboxState ? -1 : 0;
```
Comment (sroy3) [inline src/components/checkbox/GroupableCheckboxReducersUtils.ts]: You cannot edit the current state. You have to return a new one.

File: src/components/checkbox/tests/GroupableCheckboxReducers.spec.ts [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+
+        it('should have nbChecked equal to 1 on add checkbox checked with an existing state', () => {
+          action = addGroupedCheckbox('id', true, parentId);
+          const groupableCheckboxsState: IGroupableCheckboxesState[] = groupableCheckboxesReducer(currentState, action);
+          expect(groupableCheckboxsState.length).toBe(1, 'child checkbox was not added in the new state');
+          if (groupableCheckboxsState.length) {
+            expect(groupableCheckboxsState[0].nbChecked).toBe(1, 'nbChecked has been updated');
+          }
+        });
+      });
```
Comment (sroy3) [inline src/components/checkbox/tests/GroupableCheckboxReducers.spec.ts]: You should add a few tests to make sure the state being returned is a new object when it is changing.

File: src/components/checkbox/GroupableCheckboxReducersUtils.ts (hunk 9 above)
Comment (pochretien) [reply src/components/checkbox/GroupableCheckboxReducersUtils.ts]: checkboxes is already modified, the component will render and change the parent props to check, uncheck or indeterminated

File: src/components/checkbox/GroupableCheckboxConnected.tsx [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+                     groupableCheckboxesState.nbChecked < groupableCheckboxesState.total,
+    };
+  }
+
+  return {
+    checked: ownProps.checked,
```
Comment (MichelML) [inline src/components/checkbox/GroupableCheckboxConnected.tsx]: this is a bit weird that you use the ownProps.checked here, in theory it should be undefined since you are in a connected component, could this do the same

```
const mapStateToProps = (state: IReactVaporState, ownProps: IGroupableCheckboxOwnProps): IInputStateProps => {
    const groupableCheckboxesState: IGroupableCheckboxesState = _.findWhere(state.groupableCheckboxes, {parentId: ownProps.parentId || ownProps.id});
    let checkbox: ICheckboxState;
    if (groupableCheckboxesState) {
        checkbox = ownProps.isParent
            ? groupableCheckboxesState.parent
            : _.findWhere(groupableCheckboxesState.checkboxes, {id: ownProps.id});
    }

    return {
        checked: checkbox && checkbox.checked,
        indeterminate: ownProps.isParent
            && groupableCheckboxesState.nbChecked > 0
            && groupableCheckboxesState.nbChecked < groupableCheckboxesState.total,
    };
};  
```

File: src/components/checkbox/GroupableCheckboxReducers.ts (hunk 5 above)
Comment (pochretien) [reply src/components/checkbox/GroupableCheckboxReducers.ts]: because we only update the state because its already existe by a previous checkbox ( child or parent ) added. I only need to update the state with the new one.

File: src/components/checkbox/GroupableCheckboxConnected.tsx (hunk 11 above)
Comment (pochretien) [reply src/components/checkbox/GroupableCheckboxConnected.tsx]: updated, I use defaultChecked props for the first render

File: src/components/checkbox/GroupableCheckboxReducers.ts (hunk 5 above)
Comment (sroy3) [reply src/components/checkbox/GroupableCheckboxReducers.ts]: You aren't really updating the state if you are returning the old state.

File: src/components/checkbox/tests/GroupableCheckboxReducers.spec.ts (hunk 10 above)
Comment (pochretien) [reply src/components/checkbox/tests/GroupableCheckboxReducers.spec.ts]: Added

**Comment by pochretien** [comment]
I will fix the 0.04% coverage missing

File: src/components/checkbox/GroupableCheckboxConnected.tsx [hunk 12]
```diff
@@ -0,0 +1,45 @@
+import { connect } from 'react-redux';
+import * as _ from 'underscore';
+import { IReactVaporState } from '../../ReactVapor';
+import { IReduxAction, ReduxUtils } from '../../utils/ReduxUtils';
+import { IInputDispatchProps, IInputProps, IInputStateProps } from '../input/Input';
+import {Checkbox} from './Checkbox';
+import {ICheckboxState} from './CheckboxReducers';
+import {
+  addGroupedCheckbox, IGroupableCheckboxActionPayload, removeGroupedCheckbox,
+  toggleGroupedCheckbox,
```
Comment (sroy3) [inline src/components/checkbox/GroupableCheckboxConnected.tsx]: Optiize imports

File: src/components/checkbox/GroupableCheckboxReducersUtils.ts (hunk 9 above)
Comment (sroy3) [reply src/components/checkbox/GroupableCheckboxReducersUtils.ts]: You should work on another state object

File: src/components/tables/examples/TableExamples.tsx [hunk 13]
```diff
@@ -6,13 +6,13 @@ import * as _ from 'underscore';
...
 import { IDropdownOption } from '../../dropdownSearch/DropdownSearch';
-import { ITableOwnProps } from '../Table';
 import { IData, ITableRowData } from '../Table';
+import { ITableOwnProps } from '../Table';
 import { modifyState, setIsInError } from '../TableActions';
 import { TableConnected } from '../TableConnected';
 import { DEFAULT_TABLE_DATA, TABLE_PREDICATE_DEFAULT_VALUE } from '../TableConstants';
 import { defaultTableStateModifier, dispatchPostTableStateModification, dispatchPreTableStateModification } from '../TableDataModifier';
-import { ITableCompositeState, ITableData, ITablesState, ITableState } from '../TableReducers';
+import { ITableCompositeState, ITableData, ITablesState, ITableState} from '../TableReducers';
```
Comment (sroy3) [inline src/components/tables/examples/TableExamples.tsx]: There is a space at the beginning, but none at the end

File: docs/Index.tsx [hunk 14]
```diff
@@ -1,84 +1,80 @@
...
-import {TableRowConnectedExamples} from '../src/components/tables/examples/TableRowConnectedExamples';
-import {TableRowExamples} from '../src/components/tables/examples/TableRowExamples';
-import {TitleExamples} from '../src/components/title/examples/TitleExamples';
-import {ToastConnectedExamples} from '../src/components/toast/examples/ToastConnectedExamples';
-import {ToastExamples} from '../src/components/toast/examples/ToastExamples';
-import {TooltipExamples} from '../src/components/tooltip/examples/TooltipExamples';
-import {UserFeedbackExample} from '../src/components/userFeedback/examples/UserFeedbackExample';
-import {MembersExample} from './members-example/MembersExample';
-import {ReactVaporStore} from './ReactVaporStore';
+import { render as ReactDOMRender } from 'react-dom';
```
Comment (GermainBergeron) [inline docs/Index.tsx]: I think you should revert this change. We're "fighting" for the spaces in every PR which is annoying

File: src/components/checkbox/Checkbox.tsx [hunk 15]
```diff
@@ -11,6 +12,21 @@ export class Checkbox extends React.Component<IInputProps, any> {
...
+  componentDidMount() {
+    this.updateIndeterminate();
+  }
+
+  componentDidUpdate() {
+    this.updateIndeterminate();
+  }
+
+  private updateIndeterminate() {
+    const inputElements = ReactDOM.findDOMNode(this).getElementsByTagName('input');
```
Comment (GermainBergeron) [inline src/components/checkbox/Checkbox.tsx]: I'm more used to `.querySelector('.coveo-checkbox')` but I don't really mind

File: src/components/checkbox/GroupableCheckboxReducers.ts [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+          groupableCheckboxes.parentId === (action.payload.parentId || action.payload.id)
+          ? _.extend({}, groupableCheckboxes, groupableCheckboxReducer(groupableCheckboxes, action))
+          : groupableCheckboxes,
+        );
+      }
+
+      return [
+        ...state,
+        groupableCheckboxReducer(undefined, action),
+      ];
```
Comment (GermainBergeron) [inline src/components/checkbox/GroupableCheckboxReducers.ts]: We talked a bit about this yesterday and I came up with this idea:

```
let exists = false;
const newState = _.map(state, (groupableCheckboxes: IGroupableCheckboxesState) => {
    const shouldModify = groupableCheckboxes.parentId === (action.payload.parentId || action.payload.id);
    exists = shouldModify || exists;

   return shouldModify
    ? _.extend({}, groupableCheckboxes, groupableCheckboxReducer(groupableCheckboxes, action))
    : groupableCheckboxes;
});

return exists 
    ? newState 
    : [
        ...state,
        groupableCheckboxReducer(undefined, action),
      ];
```

File: src/components/checkbox/GroupableCheckboxReducersUtils.ts [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+      isChecked = !checkbox.checked;
+      return {
+        id: checkbox.id,
+        checked: isChecked,
+      };
+    }
+    return checkbox;
+  });
+  if (checkboxes) {
+    const nbChecked: number = state.nbChecked += (isChecked ? 1 : -1);
```
Comment (GermainBergeron) [inline src/components/checkbox/GroupableCheckboxReducersUtils.ts]: you could probably do `const nbChecked = _.where(state.checkboxes, {checked: true}).length;` to be sure to always have the correct number

File: docs/Index.tsx (hunk 14 above)
Comment (MichelML) [reply docs/Index.tsx]: After merging with master and rebuilding the project, if you run `npm run lintfix` it should fix all this.

File: src/components/checkbox/Checkbox.tsx [hunk 18]
```diff
@@ -35,6 +36,17 @@ export class Checkbox extends React.Component<ICheckboxProps, any> {
         }
     }
 
+    private handleOnClick(e: React.MouseEvent<HTMLElement>) {
+        if (!!!this.props.disabled) {
```
Comment (sroy3) [inline src/components/checkbox/Checkbox.tsx]: 3 `!` does this have a special meaning? If you only meant not disabled, only one `!` will do.

File: src/components/checkbox/GroupableCheckboxReducers.ts [hunk 19]
```diff
@@ -4,7 +4,7 @@ import {ICheckboxState} from './CheckboxReducers';
 import {GroupableCheckboxActions, IGroupableCheckboxActionPayload} from './GroupableCheckboxActions';
 import {
     addChildCheckbox,
-    addParentCheckbox,
+    addParentCheckbox, disabledAllCheckbox, disabledChildCheckbox, disabledParentCheckbox,
```
Comment (sroy3) [inline src/components/checkbox/GroupableCheckboxReducers.ts]: Optimize imports

File: src/components/checkbox/GroupableCheckboxReducers.ts [hunk 20]
```diff
@@ -35,6 +35,16 @@ export const groupableCheckboxReducer = (state: IGroupableCheckboxesState = grou
...
+        case GroupableCheckboxActions.disabledGroup:
+            if (state.parentId === action.payload.id || state.parentId === action.payload.parentId) {
+                return action.payload.isParent
+                    ? disabledParentCheckbox(state, action)
+                    : disabledChildCheckbox(state, action);
+            }
+        case GroupableCheckboxActions.disabledAllGroup:
+            if (state.parentId === action.payload.id || state.parentId === action.payload.parentId) {
+                return disabledAllCheckbox(state, action);
+            }
```
Comment (sroy3) [inline src/components/checkbox/GroupableCheckboxReducers.ts]: I really like your solution to simply end up in default instead of always have an else and return the same thing in each case.

File: src/components/checkbox/GroupableCheckboxReducersUtils.ts [hunk 21]
```diff
@@ -93,16 +99,73 @@ export const toggleChildCheckbox = (state: IGroupableCheckboxesState, action: IR
...
+            disabled: isDisabled,
+        },
+    };
+};
+
+export const disabledChildCheckbox = (state: IGroupableCheckboxesState, action: IReduxAction<IGroupableCheckboxActionPayload>) => {
+    let isDisabled: boolean;
+    const checkboxes: ICheckboxState[] = _.map(state.checkboxes, (checkbox: ICheckboxState) => {
+        if (checkbox.id === action.payload.id) {
+            isDisabled = !checkbox.disabled;
```
Comment (sroy3) [inline src/components/checkbox/GroupableCheckboxReducersUtils.ts]: Is the variable useful elsewhere than a few lines below? You could simply set `disabled: !checkbox.disabled` in the return.

File: src/components/checkbox/tests/Checkbox.spec.tsx [hunk 22]
```diff
@@ -30,29 +30,71 @@ describe('Checkbox', () => {
...
+                const clickSpy = jasmine.createSpy('onClick');
+                renderCheckbox({
+                    onClick: clickSpy,
+                    disabled: true,
+                });
+                const innerLabel = checkbox.find('div');
+
+                innerLabel.simulate('click');
+
+                expect(clickSpy.calls.count()).toBe(0);
```
Comment (sroy3) [inline src/components/checkbox/tests/Checkbox.spec.tsx]: `expect(clickSpy).not.toHaveBeenCalled()`

File: src/components/checkbox/tests/Checkbox.spec.tsx [hunk 23]
```diff
@@ -30,29 +30,71 @@ describe('Checkbox', () => {
...
+                const handleOnClickSpy = jasmine.createSpy('handleOnClick');
+                renderCheckbox({
+                    handleOnClick: handleOnClickSpy,
+                });
+
+                const innerLabel = checkbox.find('div');
+
+                innerLabel.simulate('click');
+
+                expect(handleOnClickSpy.calls.count()).toBe(1);
```
Comment (sroy3) [inline src/components/checkbox/tests/Checkbox.spec.tsx]: `expect(handleOnClickSpy).toHaveBeenCalledTimes(1)`

File: src/components/checkbox/Checkbox.tsx [hunk 24]
```diff
@@ -11,6 +21,32 @@ export class Checkbox extends React.Component<IInputProps, any> {
...
+
+    private updateIndeterminate() {
+        const inputElements = ReactDOM.findDOMNode(this).getElementsByTagName('input');
+        if (inputElements.length) {
+            inputElements[0].indeterminate = !!this.props.indeterminate;
+        }
+    }
+
+    private handleOnClick(e: React.MouseEvent<HTMLElement>) {
+        if (!!!this.props.disabled) {
```
Comment (MichelML) [inline src/components/checkbox/Checkbox.tsx]: `!this.props.disabled` is the same as a triple exlamation

File: src/components/checkbox/GroupableCheckboxReducers.ts (hunk 20 above)
Comment (MichelML) [reply src/components/checkbox/GroupableCheckboxReducers.ts]: I agree this is very clean

File: src/components/checkbox/GroupableCheckboxReducersUtils.ts [hunk 25]
```diff
@@ -0,0 +1,171 @@
...
+    const newCheckbox = {
+        id: action.payload.id,
+        checked: !!action.payload.checked,
+        disabled: !!action.payload.disabled,
+    };
+
+    if (state === groupableCheckboxInitialState) {
+        return {
+            ...state,
+            parentId: action.payload.id,
```
Comment (MichelML) [inline src/components/checkbox/GroupableCheckboxReducersUtils.ts]: thinking outloud, do you absolutely need `parentId` if you have it thourgh `parent.id` ?

**Comment by MichelML** [comment]
https://coveo.github.io/react-vapor/add-indeterminate-checkbox-state-and-groupable-checkboxes/ demo

File: src/components/checkbox/Checkbox.tsx (hunk 18 above)
Comment (pochretien) [reply src/components/checkbox/Checkbox.tsx]: I can receive an undefined disabled, with this 3!, I'm sure I will have a condition boolean

File: src/components/checkbox/GroupableCheckboxReducersUtils.ts (hunk 25 above)
Comment (pochretien) [reply src/components/checkbox/GroupableCheckboxReducersUtils.ts]: this variable have 2 functionalities, give an id for the groupableCheckbox and to handle the problem. When its time to add each checkbox in the state.,I'm not sure the parent will be added first. I created this variable to simplify the process for the add.

File: src/components/checkbox/Checkbox.tsx (hunk 18 above)
Comment (sroy3) [reply src/components/checkbox/Checkbox.tsx]: It does not matter, it's an if `!this.props.disabled` and `!!!this.props.disabled` will always have the exact same result here.
